src/node/convolver.rs

```diff
@@ -0,0 +1,573 @@
+/// convolve.set_buffer(Some(impulse_buffer));
+///
+/// src.connect(&convolve);
+/// convolve.connect(&context.destination());
+/// src.start();
+/// std::thread::sleep(std::time::Duration::from_millis(4_000));
+/// ```
+///
+/// # Examples
+///
+/// - `cargo run --release --example convolution`
+///
+pub struct ConvolverNode {
+    /// Represents the node instance and its associated audio context
+    registration: AudioContextRegistration,
+    /// Info about audio node channel configuration
+    channel_config: ChannelConfig,
+    /// Perform equal power normalization on response buffer
+    normalize: bool,
+    /// The response buffer, nullable
+    buffer: Mutex<Option<AudioBuffer>>,
+    /// Message bus to the renderer
+    sender: Sender<ConvolverRendererInner>,
+}
+
+impl AudioNode for ConvolverNode {
+    fn registration(&self) -> &AudioContextRegistration {
+        &self.registration
+    }
+
+    fn channel_config(&self) -> &ChannelConfig {
+        &self.channel_config
+    }
+
+    fn number_of_inputs(&self) -> usize {
+        1
+    }
+
+    fn number_of_outputs(&self) -> usize {
+        1
+    }
+}
+
+impl ConvolverNode {
+    /// returns a `ConvolverNode` instance
+    ///
+    /// # Arguments
+    ///
+    /// * `context` - audio context in which the audio node will live.
+    /// * `options` - convolver options
+    ///
+    /// # Panics
+    ///
+    /// Panics when an AudioBuffer is provided via the `ConvolverOptions` with a sample rate
+    /// different from the audio context sample rate.
+    pub fn new<C: BaseAudioContext>(context: &C, options: ConvolverOptions) -> Self {
+        context.base().register(move |registration| {
+            let ConvolverOptions {
+                buffer,
+                disable_normalization,
+                channel_config,
+            } = options;
+
+            // Channel to send buffer channels references to the renderer.  A capacity of 1
+            // suffices, it will simply block the control thread when used concurrently
+            let (sender, receiver) = crossbeam_channel::bounded(1);
+
+            let renderer = ConvolverRenderer::new(receiver);
+
+            let node = Self {
+                registration,
+                channel_config: channel_config.into(),
+                normalize: !disable_normalization,
+                sender,
+                buffer: Mutex::new(None),
+            };
+
+            node.set_buffer(buffer);
+
+            (node, Box::new(renderer))
+        })
+    }
+
+    /// Get the current impulse response buffer
+    #[allow(clippy::missing_panics_doc)]
+    pub fn buffer(&self) -> Option<AudioBuffer> {
+        self.buffer.lock().unwrap().clone()
+    }
+
+    /// Set or update the impulse response buffer
+    ///
+    /// # Panics
+    ///
+    /// Panics when the sample rate of the provided AudioBuffer differs from the audio context
+    /// sample rate.
+    pub fn set_buffer(&self, mut buffer: Option<AudioBuffer>) {
```

Comment:
For consistency, maybe we should have the same API here as for the `AudioScheduledSourceNode`

https://github.com/orottier/web-audio-api-rs/blob/d003e9938b056f0d840b8f832bd21510e7fb3ddd/src/node/audio_buffer_source.rs#L250

In both case this is `buffer, of type [AudioBuffer](https://webaudio.github.io/web-audio-api/#AudioBuffer), nullable` in the spec. 
For code readability I personally prefer the `pub fn set_buffer(&self, buffer: AudioBuffer)` version, but for spec compliance the `Option` is probably a better choice.

Reply:
Good point, but I am not sure what is best.
The ConvolverNode is different because you can set the buffer multiple times. This is not allowed for AudioBufferSourceNode.
Hence I found it necessary to allow the nullable option here. To disable convolution temporarily and add in a new buffer later.
But yeah, this is inconsistent. And probably no-one in their right mind would reuse a convolvernode for different impulse responses. I will make the API consistent with AudioBufferSourceNode